File: akit/src/main/java/org/littletonrobotics/junction/inputs/LoggedPowerDistribution.java
// Copyright (c) 2021-2025 Littleton Robotics
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by a BSD
// license that can be found in the LICENSE file
// at the root directory of this project.

package org.littletonrobotics.junction.inputs;

import edu.wpi.first.wpilibj.PowerDistribution;
import org.littletonrobotics.conduit.ConduitApi;
import org.littletonrobotics.junction.LogTable;

/** Manages logging power distribution data. */
public class LoggedPowerDistribution {

  private static LoggedPowerDistribution instance;

  private int busID;
  private int moduleID;
  private int moduleType;

  private LoggedPowerDistribution(
      int busID, int moduleID, PowerDistribution.ModuleType moduleType) {
    this.busID = busID;
    this.moduleID = moduleID;
    this.moduleType = moduleType.value;
    ConduitApi.getInstance().configurePowerDistribution(busID, moduleID, this.moduleType);
  }

  public static LoggedPowerDistribution getInstance() {
    return instance;
  }

  public static LoggedPowerDistribution getInstance(
      int busID, int moduleID, PowerDistribution.ModuleType moduleType) {
    if (instance == null) {
      instance = new LoggedPowerDistribution(busID, moduleID, moduleType);
    } else if (instance.busID != busID
        || instance.moduleID != moduleID
        || instance.moduleType != moduleType.value) {
      instance = new LoggedPowerDistribution(busID, moduleID, moduleType);
    }

    return instance;
  }

  public void saveToLog(LogTable table) {
    ConduitApi conduit = ConduitApi.getInstance();
    table.put("Temperature", conduit.getPDPTemperature());
    table.put("Voltage", conduit.getPDPVoltage());
    double[] currents = new double[24];
    for (int i = 0; i < 24; i++) {
      currents[i] = conduit.getPDPChannelCurrent(i);
    }
    table.put("ChannelCurrent", currents);
    table.put("TotalCurrent", conduit.getPDPTotalCurrent());
    table.put("TotalPower", conduit.getPDPTotalPower());
    table.put("TotalEnergy", conduit.getPDPTotalEnergy());

    table.put("ChannelCount", conduit.getPDPChannelCount());
    table.put("Faults", conduit.getPDPFaults());
    table.put("StickyFaults", conduit.getPDPStickyFaults());
  }
}

File: akit/src/main/native/cpp/LoggedPowerDistribution.cpp
// Copyright (c) 2021-2026 Littleton Robotics
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by a BSD
// license that can be found in the LICENSE file
// at the root directory of this project.

#include "akit/LoggedPowerDistribution.h"
#include "akit/conduit/ConduitApi.h"

using namespace akit;

std::unique_ptr<LoggedPowerDistribution> LoggedPowerDistribution::instance;

LoggedPowerDistribution& LoggedPowerDistribution::GetInstance() {
	if (!instance)
		instance = std::make_unique<LoggedPowerDistribution>();
	return *instance;
}

LoggedPowerDistribution& LoggedPowerDistribution::GetInstance(int moduleID,
		HAL_PowerDistributionType moduleType) {
	if (!instance)
		instance = std::make_unique < LoggedPowerDistribution
				> (moduleID, moduleType);
	return *instance;
}

void LoggedPowerDistribution::SaveToLog(LogTable &&table) {
	conduit::ConduitApi &inst = conduit::ConduitApi::getInstance();
	table.Put("Temperature", inst.getPDPTemperature());
	table.Put("Voltage", inst.getPDPVoltage());
	std::array<double, 24> currents;
	for (size_t i = 0; i < currents.size(); i++)
		currents[i] = inst.getPDPChannelCurrent(i);
	table.Put("ChannelCurrent",
			std::vector<double> { currents.begin(), currents.end() });
	table.Put("TotalCurrent", inst.getPDPTotalCurrent());
	table.Put("TotalPower", inst.getPDPTotalPower());
	table.Put("TotalEnergy", inst.getPDPTotalEnergy());

	table.Put("ChannelCount", inst.getPDPChannelCount());
	table.Put("Faults", inst.getPDPFaults());
	table.Put("StickyFaults", inst.getPDPStickyFaults());
}

LoggedPowerDistribution::LoggedPowerDistribution(int moduleID,
		HAL_PowerDistributionType moduleType) : moduleID { moduleID }, moduleType {
		moduleType } {
	conduit::ConduitApi::getInstance().configurePowerDistribution(moduleID,
			moduleType);
}
